File: DPBO-praktikum/1/Python/Main.py
# import class
from Mahasiswa import Mahasiswa
from Tabel import Tabel

n=0#jumlah data
daftar_mahasiswa = []#array daftar mahasiswa
arrheader=["NAMA", "NIM", "Prodi", "Fakultas"]#array header
max=[0] * 4#array string terpanjang
nama=""#penampung data mahasiswa
nim=""
prodi=""
fakultas=""
for i in range(4):#tentukan string terpanjang
    max[i]=max[i] if max[i]>len(arrheader[i]) else len(arrheader[i])

while True:
    print("1. Tambah Data")
    print("2. Tampilkan Daftar")
    print("3. Update Data")
    print("4. Hapus Data")
    print("5. Keluar")
    print("-----------------------------")
    pil=int(input("Masukan Pilihan(1-5) :"))#inputkan pilihan

    if pil==1:
        n=int(input("\nMasukan banyak Mahasiswa :"))#inputkan jumlah data
        for i in range(n):
            print("Masukan data mahasiswa ke",i+1)#masukan data lalu cek string sterpanjangnya
            print("Nama     : ", end="")
            nama=input()
            max[0] = max[0] if max[0] > len(nama) else len(nama)
            print("NIM      : ", end="")
            nim=input()
            max[1] = max[1] if max[1] > len(nim) else len(nim)
            print("Prodi    : ", end="")
            prodi=input()
            max[2] = max[2] if max[2] > len(prodi) else len(prodi)
            print("Fakultas : ", end="")
            fakultas=input()
            max[3] = max[3] if max[3] > len(fakultas) else len(fakultas)
            daftar_mahasiswa.append(Mahasiswa(nama, nim, prodi, fakultas))#tambahkan objek mahasiswa baru ke daftar
            print("-----------------------------")
        print(n,"data berhasil ditambahkan ke daftar\n")

    elif pil==2:
        if not daftar_mahasiswa:#jika daftar masih kosong
            print("\n=== Daftar kosong ===\n")
        else:#jika tidak
            tbl=Tabel(n, 4)#buat objek tabel
            tbl.buat_baris(arrheader, max, sum(max), '=')#tampilkan header
            for mhs in daftar_mahasiswa:
                arrisi=[ mhs.get_name(), mhs.get_nim(), mhs.get_prodi(), mhs.get_fakultas()]#isi array baris
                tbl.buat_baris(arrisi, max, sum(max), '-')#tampilkan baris

    elif pil==3:#update
        if not daftar_mahasiswa:#jika daftar kosong
            print("\n=== Daftar kosong ===\n")
        else:    
            print("Masukan NIM yang akan diupdate :")#inputkan nim data yang ingin diupdate
            nim=input()
            found=False
            for mhs in daftar_mahasiswa:#cari dalam daftar
                if mhs.get_nim() == nim:#jika ketemu
                    found=True
                    print("Nama: ", mhs.get_name())#tampilkan data awal
                    print("NIM: ", mhs.get_nim())
                    print("Program Studi: ", mhs.get_prodi())
                    print("Fakultas: ", mhs.get_fakultas())
                    print("Update Data: ")#minta inputan data baru
                    nama=input()
                    prodi=input()
                    fakultas=input()
                    max[0] = max[0] if max[0] > len(nama) else len(nama)#cek string terpanjang
                    max[2] = max[2] if max[2] > len(prodi) else len(prodi)
                    max[3] = max[3] if max[3] > len(fakultas) else len(fakultas)
                    mhs.set_name(nama)#set atribut objek
                    mhs.set_prodi(prodi)
                    mhs.set_fakultas(fakultas)
            if not found:#jika tidak ditemukan
                print("NIM tidak ditemukan!")

    elif pil==4:#delete
        if not daftar_mahasiswa:#jika daftar kosong
            print("\n=== Daftar kosong ===\n")
        else:   
            print("Masukan NIM yang akan dihapus :")#masukan nim data yang ingin dihapus
            nim=input()
            found=False
            for i, mhs in enumerate(daftar_mahasiswa):#cari dalam daftar
                if mhs.get_nim() == nim:#jika ketemu
                    found=True
                    del daftar_mahasiswa[i]#hapus objek
                    break
            if not found:#jika tidak ditemukan
                print("NIM tidak ditemukan!")

    elif pil==5:
        break

    else:
        print("Masukan pilihan yang benar!\n")
        continue

print("\n=== TERIMA KASIH ===")

File: DPBO-praktikum/1/CPP/Main.cpp
//memanggil class mahasiswa dan tabel
#include "Mahasiswa.cpp"
#include "Tabel.cpp"

int main(){
    int pil, n;//var pilihan dan jumlah input data mahasiswa
    list<Mahasiswa> daftar_Mahasiswa;//list daftar
    array<string, 4> arrheader = {"Nama", "NIM", "Prodi", "Fakultas"};//array header tabel
    array<string, 4> arrisi = { };//array isi tabel
    int max[4] = {};//array string terpanjang
    for(int i=0; i<4; i++)
        max[i] = (max[i] > arrheader[i].length()) ? max[i] : arrheader[i].length();//simpan nilai string terpanjang
    do{
        cout<<"1. Tambah Data"<<endl;
        cout<<"2. Tampilkan Daftar"<<endl;
        cout<<"3. Update Data"<<endl;
        cout<<"4. Hapus Data"<<endl;
        cout<<"5. Keluar"<<endl;
        cout<<"-----------------------------"<<endl; 
        cout<<"Masukan Pilihan(1-5) : ";cin>>pil;//inputan pilihan
        cin.ignore();//bersihkan buffer input
        switch (pil){
        case 1://tambah mahasiswa
            cout<<"\nMasukan banyak Mahasiswa : ";cin>>n;//masukan jmlh mahasiswa
            cin.ignore();//bersihkan buffer input
            for(int i=0; i<n; i++){
                cout<<"Masukan data mahasiswa ke"<<i+1<<endl;//isi data mahasiswa
                cout<<"Masukan Nama          : ";
                getline(cin, arrisi[0]);
                cout<<"Masukan NIM           : ";
                getline(cin, arrisi[1]);
                cout<<"Masukan Program Studi : ";
                getline(cin, arrisi[2]);
                cout<<"Masukan Fakultas      : ";
                getline(cin, arrisi[3]);
                for(int i=0; i<4; i++)
                    max[i] = (max[i] > arrisi[i].length()) ? max[i] : arrisi[i].length();//simpan nilai string terpanjang
                daftar_Mahasiswa.push_back(Mahasiswa(arrisi[0], arrisi[1],arrisi[2],arrisi[3]));//tambahkan ke list daftar mahasiswa
                cout<<"-----------------------------"<<endl;
            }
            cout<<n<<" data berhasil ditambahkan ke daftar"<<endl<<endl;
            break;
        case 2://update data
            cout<<"-------------------------------"<<endl;
            if (daftar_Mahasiswa.empty()) {//jika daftar masih kosong
                cout << "\n=== Daftar kosong ===\n" << endl;
            } else {//jika tidak
                Tabel tbl(daftar_Mahasiswa.size(), 4);//buat objek tabel dengan 4 kolom dan baris sebanyak isi list
                tbl.buat_baris(arrheader, max, max[0] + max[1] + max[2] + max[3], '=');//buat baris header
                for (Mahasiswa mhs : daftar_Mahasiswa) {
                    arrisi = { mhs.get_name(), mhs.get_nim(), mhs.get_prodi(), mhs.get_fakultas() };//mengisi array
                    tbl.buat_baris(arrisi, max, max[0] + max[1] + max[2] + max[3], '-');//buat baris isi
                }
            }
            cout<<endl;
            break;
        case 3://update
            cout<<"-------------------------------"<<endl;
            if (daftar_Mahasiswa.empty()) {
                cout << "\n=== Daftar kosong ===\n" << endl;
            } else {
                cout << "\nMasukkan NIM mahasiswa yang akan diupdate: ";
                getline(cin, arrisi[1]);//meminta inputan nim
                for (list<Mahasiswa>::iterator it = daftar_Mahasiswa.begin(); it != daftar_Mahasiswa.end(); it++) {//loop mencari nim
                    if (it->get_nim() == arrisi[1]) {//jika ketemu, maka masukan data baru
                        cout << "Masukkan Nama          : ";
                        getline(cin, arrisi[0]);
                        cout << "Masukkan Program Studi : ";
                        getline(cin, arrisi[2]);
                        cout << "Masukkan Fakultas      : ";
                        getline(cin, arrisi[3]);
                        for(int i=0; i<4; i++)
                            max[i] = (max[i] > arrisi[i].length()) ? max[i] : arrisi[i].length();//cek kembali string terpanjangnya
                        it->set_name(arrisi[0]);//ubah atribut dengan set
                        it->set_prodi(arrisi[2]);
                        it->set_fakultas(arrisi[3]);
                        cout << "\nData mahasiswa berhasil diupdate\n" << endl;
                        break;
                    }
                }
            }
            break;
        case 4://delete
            cout<<"-------------------------------"<<endl;
            if (daftar_Mahasiswa.empty()) {//jika daftar kosong
                cout << "\n=== Daftar kosong ===\n" << endl;
            } else {
                cout << "\nMasukkan NIM mahasiswa yang akan dihapus: ";
                getline(cin, arrisi[1]);//masukan nim
                for (list<Mahasiswa>::iterator it = daftar_Mahasiswa.begin(); it != daftar_Mahasiswa.end(); it++) {//cari nim
                    if (it->get_nim() == arrisi[1]) {//jika ketemu
                        daftar_Mahasiswa.erase(it);//hapus dari daftar
                        cout << "\nData mahasiswa berhasil dihapus\n" << endl;
                        break;
                    }
                }
            }
            break;
        case 5://out
            break;    
        default:
            cout<<"-------------------------------"<<endl;
            cout << "\n=== Masukan Input dengan Benar! ===\n" << endl;
            break;
        }
    }while(pil!=5);
    cout<<"\n=== TERIMA KASIH ==="<<endl;
}
